content/blog/2018-06-07-when-to-use-derived-state.md

```diff
@@ -0,0 +1,331 @@
+    // Re-run the filter whenever the list array or filter text change:
+    if (
+      props.list !== state.prevPropsList ||
+      state.prevFilterText !== state.filterText
+    ) {
+      // Note that to do this, state needs to explicitly track both:
+      // prevPropsList - so we can detect a change in props.list
+      // prevFilterText - so we can detect a change in state.filterText
+      return {
+        prevPropsList: props.list,
+        prevFilterText: state.filterText,
+        filteredList: props.list.filter(item => item.text.includes(state.filterText))
+      };
+    }
+    return null;
+  }
+
+  render() {
+    return (
+      <Fragment>
+        <input onChange={this.handleChange} value={this.state.filterText} />
+        <ul>{this.state.filteredList.map(item => <li key={item.id}>{item.text}</li>)}</ul>
+      </Fragment>
+    );
+  }
+
+  handleChange = event => this.setState({ filterText: event.target.value });
+}
+```
+
+This implementation is more complicated than it needs to be, because it has to separately track and detect changes in both props and state in order to properly memoize. This is true of any memoized function, which suggests an alternative! We could accomplish the behavior with **less complexity** using a memoization helper:
+
+```js
+import memoize from "memoize-one";
+
+class Example extends Component {
+  // State only needs to hold the current filter text value:
+  state = {
+    filterText: ""
+  };
+
+  // Re-run the filter whenever the list array or filter text change:
+  filter = memoize(
+    (list, filterText) => list.filter(
+      item => item.text.includes(filterText)
+    )
+  );
+
+  render() {
+    const filteredList = this.filter(this.props.list, this.state.filterText);
+
+    return (
+      <Fragment>
+        <input onChange={this.handleChange} value={this.state.filterText} />
+        <ul>{filteredList.map(item => <li key={item.id}>{item.text}</li>)}</ul>
+      </Fragment>
+    );
+  }
+
+  handleChange = event => this.setState({ filterText: event.target.value });
+}
+```
+
+This is much simpler and should perform just as well as the previous version!
+
+There are a couple of constraints to consider when using memoization though:
+
+1. In most cases, you'll want to **attach the memoized function to a component instance**. This prevents multiple instances of a component from resetting each other's memoized keys.
+1. Typically you'll want to use a memoization helper with a **limited cache size** in order to prevent memory from "leaking" over time. (In the example above, we used `memoize-one` because it only caches the most recent argument and result.)
+1. Memoization will not work properly if `props.items` is recreated each time the parent component renders. Fortunately, this should not be the case for large arrays. If it were, you'd want to address that problem first! (This limitation also applies to the `getDerivedStateFromProps` version above.)
```

Comment:
"Fortunately" sounds like we guarantee it somehow. "Hopefully"?

Reply:
Oh nice. I was uncertain about this wording choice to begin with. I like that.